src/metabase/models/card.clj

```diff
@@ -221,7 +237,8 @@
       (field-values/update-field-values-for-on-demand-dbs! field-ids))))
 
 (defn- pre-update [{archived? :archived, :as card}]
-  (u/prog1 card
+  ;; save the updated Card with updated read permissions.
+  (u/prog1 (assoc card :read_permissions (perms-objects-set* card :read))
```

Comment:
Roughly how big are these expected to be?

Reply:
For 99% of them it's just a single entry that looks something like

```clojure
#{"/db/1/schema/PUBLIC/table/2"}
```

for ones that reference other tables there will be additional entries (one for each table) so realistically a Card would probably have at most 3 or 4.